rust/lance/src/dataset/builder.rs

```diff
@@ -300,31 +309,96 @@ impl DatasetBuilder {
             )),
         };
 
-        let mut version: Option<u64> = None;
-        let cloned_ref = self.version.clone();
+        let target_ref = self.version.clone();
         let table_uri = self.table_uri.clone();
 
         // How do we detect which version scheme is in use?
-
         let manifest = self.manifest.take();
 
         let file_reader_options = self.file_reader_options.clone();
         let (object_store, base_path, commit_handler) = self.build_object_store().await?;
 
-        if let Some(r) = cloned_ref {
-            version = match r {
-                Ref::Version(v) => Some(v),
-                Ref::Tag(t) => {
-                    let tags = Tags::new(
-                        object_store.clone(),
-                        commit_handler.clone(),
-                        base_path.clone(),
-                    );
-                    Some(tags.get_version(t.as_str()).await?)
+        // Two cases that need to check out after loading the manifest:
+        // 1. If the target is a configured as a branch, we need to check the branch field in the manifest
+        // and reload the right branch in case the uri is not the right one.
```

Comment:
I think we should not allow publicly load a URI that is not a root table URI, user must specify the dataset using its root URI, and then specify a branch or tag to checkout. That should also avoid the case of "we don't find the tag under the table_uri"

Reply:
I think it's hard to abandon users to use a branch uri since branch could be considered as an independent dataset.

I wonder if you mean we should not introduce a double loading if we can't find the TagContent directly from the uri?

Reply:
I think these are two separate issues.

Let's say we have dataset at 's3://bucket/main', we have branches 'feature/A' and 'feature/B'.

We can checkout branch 'feature/A' by(which is our common sense):
1. load dataset at  's3://bucket/main'
2. checkout_branch or checkout_version 'feature/A'

The other way is:
1. load branch at 's3://bucket/main/feature/A'

This approach will directly get the target branch and we could know this is a branch by the 'branch' field in manifest., which means:
1. There's no difference at format layer with the first loading dataset 
2. When we create branch on this loaded dataset, we will know the root location and create the branch at the right location.
3. This approach won't need double loading.

Reply:
> The other way is:
> load branch at 's3://bucket/main/feature/A'
> This approach will directly get the target branch and we could know this is a branch by the 'branch' field in manifest.,

oh I see. i guess I misunderstood. as long as we are in the same state if we load main and then checkout, vs we load the branch directly, I think we are good